typing/typetexp.ml

```diff
@@ -953,14 +969,23 @@ let report_error env ppf = function
   | Unbound_cltype lid ->
       fprintf ppf "Unbound class type %a" longident lid;
       spellcheck ppf fold_cltypes env lid;
-  | Ill_typed_functor_application lid ->
-      fprintf ppf "Ill-typed functor application %a" longident lid
+  | Ill_typed_functor_application (lid, details) ->
+     fprintf ppf "@[Ill-typed functor application %a%t@]"
+       longident lid
+       (fun ppf -> may (fprintf ppf "@\n%a" Includemod.report_error) details)
   | Illegal_reference_to_recursive_module ->
-      fprintf ppf "Illegal recursive module reference"
-  | Access_functor_as_structure lid ->
-      fprintf ppf "The module %a is a functor, not a structure" longident lid
-  | Apply_structure_as_functor lid ->
-      fprintf ppf "The module %a is a structure, not a functor" longident lid
+     fprintf ppf "Illegal recursive module reference"
+  | Wrong_use_of_module (lid, details) ->
+     let is, but_used_as =
+       match details with
+       | `Structure_used_as_functor -> "a structure", "a functor"
+       | `Abstract_used_as_functor -> "abstract", "a functor"
+       | `Functor_used_as_structure -> "a functor", "a structure"
+       | `Abstract_used_as_structure -> "abstract", "a structure"
+       | `Generative_used_as_applicative ->
+          "a generative functor", "an applicative functor"
+     in
+     fprintf ppf "@[The module %a is %s,@ not %s@]" longident lid is but_used_as
```

Comment:
I share the opinion of @Drup, these error messages might end un being unscrutable for people not familiar with the module system since they ends un just classifying the type of the error rather than detailing the issue at hand.

I would replace

> The module M is a structure, not a functor
> The module M is abstract, not a functor
> The module M is a functor, not a structure
> The module M is abstract, not a structure
> The module M is a generative functor, not an applicative functor

by 

> The module M cannot be applied, it is a structure not a functor
> The abstract module M cannot be applied
> As a functor, the module M does not have any components that can be accessed
> As an abstract module, the module M does not have any components that can be accessed
> The generative functor M can only be applied to ( )